### PR TITLE
Add tsconfig presets for proper use of tsc and improve `jetpack generate`

### PR DESCRIPTION
**Pull request by anomiex**
## Proposed changes:

When setting `noEmit` to false for tsc, we need to make sure that `module` and `moduleResolution` are set to "nodenext" as well. Let's create some presets that do that, and set some other related configs.
    
Also let's clean up the base config a bit:
- We were setting `forceConsistentCasingInFileNames: true` everywhere, so let's go ahead and put that in the config.
- The values we had been setting for `lib` are already implied by setting `target` to "esnext" or "nodenext", so lets drop `lib`.
    
Then let's have `lint-project-structure.sh` complain if someone is trying to set these configs manually in tsconfig.json, since it's likely they won't get them all right.

And finally, let's update `jetpack generate` to do better for js-packages
    
- Ask whether it's JS or TS, and which build system (if any) is to be used.
- Use `src/index.jsx` instead of `src/index.js` by default.
- Set up webpack and/or ts deps if needed.
- Create `tsconfig.json` if TS is selected.
- Create `webpack.config.cjs` if Webpack is selected.
- Set up build scripts for the selected build system, if any.
- Don't bother asking about phpunit tests.
- Sort the resulting `scripts` and `devDependencies`.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

pdWQjU-FO-p2
Announcement post: pdWQjU-HO-p2

## Testing instructions:

* Is CI happy?
* Try reverting some of the tsconfig changes here and see if the updated `.github/files/lint-project-structure.sh` flags them.
* Try out `jetpack generate` for js-packages with various answers to the new question. Do the results look right?
* Check the built artifacts. Other than image-guide, which loses some Promise polyfills, nothing should be changing.

### CODE REVIEW

File: tools/cli/commands/generate.js [hunk 1]
```diff
@@ -415,6 +456,110 @@ function generatePlugin( answers, pluginDir ) {
...
+		opts = {
+			typeRoots: [ './node_modules/@types/', 'src/*' ],
+			outDir: './build/',
+		};
+		switch ( answers.typescript ) {
+			case 'ts-src':
+				xtends = '\n\t"extends": "jetpack-js-tools/tsconfig.base.json",';
+				break;
+			case 'ts-webpack':
+				xtends = '\n\t"extends": "jetpack-js-tools/tsconfig.tsc-definitions.json",';
```
Comment (zinigor) [inline tools/cli/commands/generate.js]: Is this `tsc-definitions` file auto-generated somehow? I can't find it in the repo.

**Review by zinigor — COMMENTED** [review]
The changes look good to me, but I'm running into a problem when testing the generator.

My choices were:

```
✔ What type of project are you working on today? · js-packages
✔ What is your project called? · bogus
✔ Succinctly describe your project: · Bugs project
✔ Does this project need to be deployed publicly? (Create a mirror repo?) (Y/n) · false
✔ Which best describes this package? · ts-webpack
```

What I got:

```
Uh oh! Cannot convert undefined or null to object
{
  type: 'js-package',
  name: 'bogus',
  _: [ 'generate' ],
  v: false,
  verbose: false,
  '$0': 'jetpack',
  n: 'bogus',
  description: 'Bugs project',
  buildScripts: [],
  wordbless: false,
  mirrorrepo: false,
  typescript: 'ts-webpack',
  project: 'js-packages/bogus',
  projDir: '/home/zinigor/workspace/jetpack/projects/js-packages/bogus'
}
```

File: tools/cli/commands/generate.js (hunk 1 above)
Comment (anomiex) [reply tools/cli/commands/generate.js]: No, it's the wrong name.

**Review by zinigor — APPROVED** [review]
Thanks for fixing, now it's working correctly for me!

**Pull request merged**